In [2]:
# Para no tener que reiniciar el kernel
#%load_ext autoreload
#%autoreload 2

# Importaciones
import pandas as pd
import numpy as np
from datetime import datetime

#Seteo de opciones
pd.set_option("display.max_columns", None)

In [3]:
import matplotlib.pyplot as plt
import seaborn as sns

### Hacemos la descarga de datos de los dos archivos (bank y customer)

In [4]:
customer_2012 = pd.read_excel("../data/data_raw/customer-details.xlsx",sheet_name = '2012', index_col=0)
customer_2013 = pd.read_excel("../data/data_raw/customer-details.xlsx",sheet_name = '2013', index_col=0)
customer_2014 = pd.read_excel("../data/data_raw/customer-details.xlsx",sheet_name = '2014', index_col=0)

In [5]:
customer_2012.shape

(20115, 6)

In [6]:
customer_2013.shape

(8965, 6)

In [7]:
customer_2014.shape

(14090, 6)

In [8]:
df_customer = pd.concat([customer_2012, customer_2013, customer_2014], axis=0, ignore_index=True)

In [25]:
df_bank = pd.read_csv("../data/data_raw/bank-additional.csv", index_col=0)

### Verificamos el tamaño de ambos df

In [10]:
print(f"El tamaño del dataset clientes es de {df_customer.shape[0]} filas y {df_customer.shape[1]} filas")

El tamaño del dataset clientes es de 43170 filas y 6 filas


In [11]:
print(f"El tamaño del dataset clientes es de {df_bank.shape[0]} filas y {df_bank.shape[1]} filas")

El tamaño del dataset clientes es de 43000 filas y 23 filas


### Estandarizamos los nombres de columnas

In [12]:
df_customer.columns = df_customer.columns.str.lower()
df_customer.columns

Index(['income', 'kidhome', 'teenhome', 'dt_customer', 'numwebvisitsmonth',
       'id'],
      dtype='object')

In [38]:
df_bank.columns = df_bank.columns.str.replace("_","").str.lower().str.replace(".","_")
df_bank.columns

Index(['age', 'job', 'marital', 'education', 'default', 'housing', 'loan',
       'contact', 'duration', 'campaign', 'pdays', 'previous', 'poutcome',
       'emp_var_rate', 'cons_price_idx', 'cons_conf_idx', 'euribor3m',
       'nr_employed', 'y', 'date', 'latitude', 'longitude', 'id'],
      dtype='object')

## Análisis inicial df_customer

In [39]:
df_customer.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 43170 entries, 0 to 43169
Data columns (total 6 columns):
 #   Column             Non-Null Count  Dtype         
---  ------             --------------  -----         
 0   income             43170 non-null  int64         
 1   kidhome            43170 non-null  int64         
 2   teenhome           43170 non-null  int64         
 3   dt_customer        43170 non-null  datetime64[ns]
 4   numwebvisitsmonth  43170 non-null  int64         
 5   id                 43170 non-null  object        
dtypes: datetime64[ns](1), int64(4), object(1)
memory usage: 2.0+ MB


In [40]:
df_customer.describe().T

count                           mean  \
income             43170.0                   93227.389622   
kidhome            43170.0                       1.004749   
teenhome           43170.0                       0.998633   
dt_customer          43170  2013-05-11 22:09:47.407922176   
numwebvisitsmonth  43170.0                       16.59129   

                                   min                  25%  \
income                          5841.0              49608.0   
kidhome                            0.0                  0.0   
teenhome                           0.0                  0.0   
dt_customer        2012-01-01 00:00:00  2012-07-16 00:00:00   
numwebvisitsmonth                  1.0                  9.0   

                                   50%                  75%  \
income                         93009.5             136740.5   
kidhome                            1.0                  2.0   
teenhome                           1.0                  2.0   
dt_customer        2013-03-01 00:00:00  2014-03-28 00:00:00   
numwebvisitsmonth                 17.0                 25.0   

                                   max           std  
income                        180802.0  50498.181989  
kidhome                            2.0      0.815996  
teenhome                           2.0      0.816008  
dt_customer        2014-12-31 00:00:00           NaN  
numwebvisitsmonth                 32.0      9.241769

In [41]:
df_customer.describe(include='O')

id
count                                  43170
unique                                 43170
top     089b39d8-e4d0-461b-87d4-814d71e0e079
freq                                       1

In [17]:
df_customer["dt_customer"] = pd.to_datetime(df_customer["dt_customer"],format = "%m/%d/%Y" )   
df_customer.head()

income  kidhome  teenhome dt_customer  numwebvisitsmonth  \
0  161770        1         0  2012-04-04                 29   
1   85477        1         1  2012-12-30                  7   
2  147233        1         1  2012-02-02                  5   
3  121393        1         2  2012-12-21                 29   
4   63164        1         2  2012-06-20                 20   

                                     id  
0  089b39d8-e4d0-461b-87d4-814d71e0e079  
1  e9d37224-cb6f-4942-98d7-46672963d097  
2  3f9f49b5-e410-4948-bf6e-f9244f04918b  
3  9991fafb-4447-451a-8be2-b0df6098d13e  
4  eca60b76-70b6-4077-80ba-bc52e8ebb0eb

### Comprobamos presencia de nulos

In [18]:
df_customer.isnull().sum()

income               0
kidhome              0
teenhome             0
dt_customer          0
numwebvisitsmonth    0
id                   0
dtype: int64

### Comprobamos duplicados en el df y en la columna id

In [19]:
df_customer.duplicated().sum()

np.int64(0)

In [20]:
df_customer.duplicated(["id"]).sum()

np.int64(0)

In [21]:
df_customer.columns

Index(['income', 'kidhome', 'teenhome', 'dt_customer', 'numwebvisitsmonth',
       'id'],
      dtype='object')

In [22]:
df_customer["kidhome"].unique()

array([1, 0, 2])

## Preanálisis df_bank

### info()

In [42]:
df_bank.sample(3)

age        job   marital    education  default  housing  loan  \
23742  31.0     admin.  DIVORCED  high.school      0.0      0.0   0.0   
6360   52.0   services   MARRIED  high.school      0.0      0.0   0.0   
27285   NaN  housemaid   MARRIED     basic.4y      0.0      1.0   0.0   

         contact  duration  campaign  pdays  previous     poutcome  \
23742   cellular        12        10    999         0  NONEXISTENT   
6360   telephone       145         5    999         0  NONEXISTENT   
27285   cellular        66         1    999         0  NONEXISTENT   

       emp_var_rate cons_price_idx cons_conf_idx euribor3m nr_employed   y  \
23742           1.4         93,444         -36,1     4,962      5228,1  no   
6360            1.1         93,994         -36,4     4,857        5191  no   
27285          -0.1           93,2           -42     4,021      5195,8  no   

                    date  latitude  longitude  \
23742  26-diciembre-2019    45.660    -74.464   
6360    8-diciembre-2017    35.335    -87.581   
27285  13-diciembre-2016    29.008    -84.830   

                                         id  
23742  c97c1826-34a5-40bc-810e-436c47cf5647  
6360   e479e5a4-b24c-497c-af1f-b16c4fc58323  
27285  d8232347-7f48-41d2-898e-a051076851b2

In [43]:
df_bank.info()

<class 'pandas.core.frame.DataFrame'>
Index: 43000 entries, 0 to 27570
Data columns (total 23 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   age             37880 non-null  float64
 1   job             42655 non-null  object 
 2   marital         42915 non-null  object 
 3   education       41193 non-null  object 
 4   default         34019 non-null  float64
 5   housing         41974 non-null  float64
 6   loan            41974 non-null  float64
 7   contact         43000 non-null  object 
 8   duration        43000 non-null  int64  
 9   campaign        43000 non-null  int64  
 10  pdays           43000 non-null  int64  
 11  previous        43000 non-null  int64  
 12  poutcome        43000 non-null  object 
 13  emp_var_rate    43000 non-null  float64
 14  cons_price_idx  42529 non-null  object 
 15  cons_conf_idx   43000 non-null  object 
 16  euribor3m       33744 non-null  object 
 17  nr_employed     43000 non-null  obje

### Conclusiones .info()

- Tenemos varias columas con nulos:
  - age
  - job
  - marital
  - education
  - default
  - housing
  - loan
  - cons.price.idx
  - euribor3m
  - date

- Columnas en las que habría que revisar/cambiar el tipo de dato
  - cons.price.idx -> object, coma en los decimales
  - cons.conf.idx -> object, coma en los decimales
  - euribor3m -> object, coma en los decimales
  - nr.employed -> object, coma en los decimales
  - date -> object, formato con el mes en texto

In [45]:
df_bank.describe().T

count        mean         std      min        25%       50%  \
age           37880.0   39.977112   10.437957   17.000   32.00000   38.0000   
default       34019.0    0.000088    0.009390    0.000    0.00000    0.0000   
housing       41974.0    0.535998    0.498708    0.000    0.00000    1.0000   
loan          41974.0    0.155620    0.362499    0.000    0.00000    0.0000   
duration      43000.0  257.739279  258.666033    0.000  102.00000  179.0000   
campaign      43000.0    2.567233    2.772294    1.000    1.00000    2.0000   
pdays         43000.0  962.330953  187.260394    0.000  999.00000  999.0000   
previous      43000.0    0.174023    0.497366    0.000    0.00000    0.0000   
emp_var_rate  43000.0    0.077128    1.573898   -3.400   -1.80000    1.1000   
latitude      43000.0   36.856697    7.225948   24.396   30.61475   36.7610   
longitude     43000.0  -95.939067   16.752282 -124.997 -110.49425  -95.8995   

                    75%       max  
age            47.00000    98.000  
default         0.00000     1.000  
housing         1.00000     1.000  
loan            0.00000     1.000  
duration      319.00000  4918.000  
campaign        3.00000    56.000  
pdays         999.00000   999.000  
previous        0.00000     7.000  
emp_var_rate    1.40000     1.400  
latitude       43.11325    49.384  
longitude     -81.42775   -66.937

### Observaciones .describe()

- Varias columnas con sólo valores 0 y 1 (default, housing, load) tienen tipo float y parecen ser booleanos
- columna pdays con valores maximos que no parecen coherentes, 999 dias como maximo no tiene sentido, revisar

In [44]:
df_bank.describe(include='O').T

count unique                                   top   freq
job             42655     11                                admin.  10873
marital         42915      3                               MARRIED  25999
education       41193      7                     university.degree  12722
contact         43000      2                              cellular  27396
poutcome        43000      3                           NONEXISTENT  37103
cons_price_idx  42529     26                                93,994   7938
cons_conf_idx   43000     26                                 -36,4   8020
euribor3m       33744    309                                 4,857   2287
nr_employed     43000     11                                5228,1  16980
y               43000      2                                    no  38156
date            42752   1825                       28-febrero-2018     95
id              43000  43000  089b39d8-e4d0-461b-87d4-814d71e0e079      1

In [ ]:
df_nulos = round(df_bank.isnull().sum() / len(df_bank) * 100, 2)
df_nulos

### Valores unicos de df_bank

In [ ]:
bank_columnas = df_bank.columns.tolist()
bank_columnas

In [48]:
for col in bank_columnas:
     print(f"Columna {col.upper()}:\n")
     print(f"Valores únicos: {df_bank[col].unique()}\n")
     print(f"Conteo de valores únicos:\n{df_bank[col].value_counts()}")
     print("_____________________________________")

Columna AGE:

Valores únicos: [nan 57. 37. 40. 56. 45. 59. 24. 25. 41. 29. 35. 54. 50. 39. 30. 55. 46.
 49. 34. 52. 58. 32. 38. 44. 42. 60. 51. 48. 47. 31. 33. 53. 43. 36. 28.
 27. 26. 22. 20. 23. 21. 61. 19. 18. 70. 66. 76. 67. 73. 88. 95. 68. 75.
 63. 80. 62. 72. 82. 64. 71. 69. 78. 65. 85. 79. 77. 83. 81. 74. 17. 87.
 91. 86. 98. 94. 84. 92. 89.]

Conteo de valores únicos:
age
31.0    1810
33.0    1704
32.0    1672
36.0    1649
34.0    1608
        ... 
91.0       2
98.0       2
95.0       1
87.0       1
94.0       1
Name: count, Length: 78, dtype: int64
_____________________________________
Columna JOB:

Valores únicos: ['housemaid' 'services' 'admin.' 'blue-collar' 'technician' 'retired'
 'management' 'unemployed' 'self-employed' nan 'entrepreneur' 'student']

Conteo de valores únicos:
job
admin.           10873
blue-collar       9654
technician        7026
services          4162
management        3050
retired           1790
entrepreneur      1522
self-employed     1489
housemaid 

### Observaciones valores únicos y nulos

- **age**
  - Las 5 edades mas repetidas son 31, 32, 33, 34 y 36 sin grandes diferencias entre ellas
  - Hay un 11.91% de nulos
- **job**
  - admin. con 10873 y blue-collar con 9654 diferenciados del resto
  - 0.80% de nulos
- **marital**
  - 4 valores únicos ('MARRIED' 'SINGLE' 'DIVORCED' nan)
  - MARRIED el más destacado, dobla en cantidad al siguiente
  - 0.20% de nulos
- **education**
  - 7 valores únicos (university.degree, high.school, basic.9y, professional.course, basic.4y, basic.6y, illiterate)
  - university.degree el más destacado con 12722 registros
  - illiterate solo tiene 18 registros?
  - 4.20% de nulos
- **default**
  - 3 valores únicos (0. nan  1.)
  - 0 valor principal, 1 solo 3 registros
  - 20.89% de nulos
- **housing**
  - 3 valores únicos (0. nan  1.)
  - poca diferencia entre valores 0 y 1
  - 2.39% de nulos
- **loan**
  - 3 valores únicos (0. nan  1.)
  - 0 con 29000 registros mas que 1
  - 2.39% de nulos 
- **contact**
  - 2 valores únicos ('telephone' 'cellular')
  - mayor presencia de cellular, muy diferenciado de relephone
- **dutation**
  - 5 valores mas destacados (90, 136, 85, 73, 111)
  -con registros entre 170 y 183
- **campaign**
  - los valores mas destacados son 1 y 2 contactos
  - bastante diferenciados de los siguientes valores
- **pdays**
  - valor más destacado 999 (no tiene sentido) con 41412 registros
- **previous**
  - valor de contactos más destacado es 0 con 37103
  - siguiente es 1 con 4778
- **poutcome**
  - 3 valores únicos ('NONEXISTENT' 'FAILURE' 'SUCCESS')
  - destaca NONEXISTENT con 16980
- **emp_var_rate**
  - 10 valores únicos
  - el más destacado es 1.4 con 16980 registros
- **cons_price_idx** 
  - sin valores que destaquen 
  - 1.10 % de nulos
- **cons_conf_idx** 
  - sin valores que destaquen
- **euribor3m**
  - sin valores que destaquen
  - 21.53% de nulos
- **nr_employed**
  - parece ser número de empleado, no numero de empleados
  - destaca notablemente 5228.1 (no tiene sentido los decimales)
- **y**
  - 2 valores unicos, no y yes
  - destaca no con 36156 registros   
- **date**
  - sin valores que destaquen
  - 0.58% de nulos
- **latitude**, **longitude** 
  - som valores significativos
- **id**
  - sin nulos ni valores que se repitan más de una vez


### Comprobacion de duplicados

In [50]:
df_bank.duplicated().sum()

np.int64(0)

Sin presencia de duplicados
